src/edu/ucsb/nceas/metacat/admin/upgrade/HashStoreUpgrader.java

```diff
@@ -185,25 +186,22 @@ public boolean upgrade() throws AdminException {
                                                     savingChecksumTableWriter);
                                         });
                                         futures.add(future);
-                                        index++;
-                                        if (index == maxListLength) {
-                                            //When it reaches the max length, we need to check
-                                            // the status in the list, then reset the list
-                                            checkFuturesStatus(futures);
-                                            futures = new ArrayList<>();
-                                            logMetacat.debug("Metacat reset the future list after"
-                                                                 + " checking futures' status");
```

Comment:
A suggestion to insert here. The logic is like this:

1. we start with no limit on the Set of futures
2. each time we add a future, we check to see how much memory is available
3. when the available memory drops below a certain level (say 20% of all available memory), then we use the current Set size as the upper limit
4. After this we stop checking the memory, and just use the upper limit, like you do now.

This is how it looks:

```java
// line 66 - change to:
    private int maxListLength = -1;


// ...then from line 188 onwards:
    futures.add(future);

    if (maxListLength < 0) {
        maxListLength = calculateMaxListLength(futures);
    }

    if (maxListLength > 0 && futures.size() >= maxListLength) {
        //When it reaches the max length, we need to remove
        // the complete futures from the set. So the set can
        // be reused again.
        removeCompleteFuture(futures);
    }
```

...where

```java
    private int calculateMaxListLength(Set<Future> futures) {

        // As soon as available memory becomes less than 20% of total
        // memory, use the current Set size as the upper limit
        if (Runtime.getRuntime().freeMemory() < 0.2 * Runtime.getRuntime().totalMemory()) {
            maxListLength = futures.size();
            return maxListLength;

        } else if (futures.size() > SAFETY_LIMIT) {
            //just in case :-) 
            return SAFETY_LIMIT;
        }
        return -1;
    }
```

obviously, it doesn't have to be `0.2` (20%) - could be anything you want

What do you think?